### PR TITLE
Retry moving mountpoint sane amount of times

### PR DESCRIPTION
**Pull request by VicDeo**
# Description
- Catch and handle integrity constraint violations and return `false` as expected
- Make 10 attempts to move the mountpoint

## Related Issue
- https://github.com/owncloud/enterprise/issues/4094

## Motivation and Context
a newly generated mounpoint path may exist but hasn't been mount yet. An attempt to move to this path will cause breaking constraint on index `sh_external_mp` for `oc_share_external` table

## How Has This Been Tested?
no exact reproduction steps provided :no_good:

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by jvillafanez** [comment]
Is it possible to get the list of the mount points from the DB somehow? Otherwise, I agree we should make the amount of times configurable.

**Comment by VicDeo** [comment]
@jvillafanez we know nothing about DB in the context of **lib/private/Files/Config/MountProviderCollection.php**
what we have there is an array of providers.

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/27320/139/15
```
--- Failed scenarios:

    /drone/src/tests/acceptance/features/webUISharingInternalGroups1/shareWithGroups.feature:133
    /drone/src/tests/acceptance/features/webUISharingInternalGroups1/shareWithGroups.feature:150
    /drone/src/tests/acceptance/features/webUISharingInternalGroups1/shareWithGroups.feature:170
    /drone/src/tests/acceptance/features/webUISharingInternalGroups1/shareWithGroups.feature:183

32 scenarios (28 passed, 4 failed)
471 steps (458 passed, 4 failed, 9 skipped)
```

This sort of problem was reported:
```
    Then a notification should be displayed on the webUI with the text "Email notification was sent!"              # WebUIGeneralContext::aNotificationShouldBeDisplayedOnTheWebUIWithTheText()
      WebUIGeneralContext::aNotificationShouldBeDisplayedOnTheWebUIWithTheText A notification was expected to be displayed on the webUI with the text 'Email notification was sent!', but got '' instead
      Failed asserting that two strings are equal.
      --- Expected
      +++ Actual
      @@ @@
      -'Email notification was sent!'
      +''
```

I don't see how the code in this PR could break that. I will restart drone...

**Comment by phil-davis** [comment]
And this time it passed.

**Comment by micbar** [comment]
@VicDeo Can we make the amount of retries configurable?

**Comment by mmattel** [comment]
This is docs relevant (config-to-docs run) because of a change in `config/config.sample.php`.
I propose to give the description in `config/config.sample.php` a bit more text for better understanding what it is about.

**Comment by VicDeo** [comment]
@mmattel frankly speaking I just made it configurable as requested. The only explanation I can add: "if you don't know what is it about then you most likely have no need to adjust it"

**Comment by mmattel** [comment]
> if you don't know what is it about then you most likely have no need to adjust it

I completely agree. In case it gets merged, 3 months later, nobody will remember what it was about. Means, you do not only add a improved description for endusers, but also for ownCloud itself to prevent that.

**Comment by VicDeo** [comment]
@mmattel  I don't know how to explain it properly

0 There could be a collision of the l mount point name for local and remote shares. E.g. when both are named `file`
1. In this case we move the mount point using the following rule - `file (1)`, `file (2)`, etc. The name is generated by checking whether `file (n)` exists **on the file system**.
2. if the storage goes offline for the remote share it could happen that `file (n)` does NOT exist **on the file system** but it's still not possible to move the mount point as we already have it in the database. So the move operation returns `false`
3. When we have `false` after the move  operation we consider `file (n)` name is unavailable and retry the move operation with `file (n+1)`. 
4. 3 is repeated exactly `filesystem.max_mountpoint_move_attempts`  times to avoid an infinite loop.

**Comment by mmattel** [comment]
**What about:**
This config value avoids infinite loops for seldom cases where a file renaming conflict could occur.
The value defines how many unsuccessful rename attempts are allowed, if the mountpoint to a file is taken but the initiated rename failed. Change this value only under supervision of ownCloud support.

Pls check the text if valid and consistent and adopt if needed.

**Comment by VicDeo** [comment]
@mmattel what about 
```
/**
 * This config value avoids infinite loops for seldom cases where a file renaming
 * conflict between different share backends could occur.
 * The value defines how many unsuccessful mountpoint rename attempts are allowed.
 * E.g. target mountpoint name could be claimed as unused by the filesystem but
 * renaming to this target name will fail due to some other reasons like database 
 * constraints.
 * Change this value only under supervision of ownCloud support.
 */
```
Is it clear enough?

**Comment by mmattel** [comment]
This is great 👍 
One thing I would do: even when starting at the beginning of the sentence, I would write `E.g.`-->`e.g.`
Else it looks at least to my eyes awful :smile:

File: config/config.sample.php [hunk 1]
```diff
@@ -1387,6 +1387,17 @@
  */
 'filesystem_check_changes' => 0,
 
+/**
+ * This config value avoids infinite loops for seldom cases where a file renaming
```
Comment (mmattel) [inline config/config.sample.php:1391]:
```suggestion
 * Define unsuccessful mountpoint rename attempts
 * This config value avoids infinite loops for seldom cases where a file renaming
```
@VicDeo There was a "headline" missing. This headline will be automatically used as item text in the table of contents on top of the document created by the config-to-docs run. I forgot to mention this when making the first text proposal. This commit suggestion fixes it. If you agree, you can securly commit it.

**Comment by phil-davis** [comment]
I don't see a changelog entry yet?

**Review by phil-davis — COMMENTED** [review]
Looks OK to me now.
@jvillafanez review?

**Review by jvillafanez — APPROVED** [review]

**Pull request merged**